### PR TITLE
extract font width logic to new class

### LINKED ISSUE #878: worksheet class, refactor font handling from method calculate_cell_width to new class
move the font logic to a separate class, including the cache

after https://github.com/sapmentors/abap2xlsx/pull/876 is merged

moving it to a separate class is first step, before tackling https://github.com/sapmentors/abap2xlsx/issues/632

### PR DESCRIPTION
**Pull request by larshp**
font width calculation logic extracted to new class, I've kept the code 1:1 where possible so its easier to review

this also adds a simple unit test for the new method

closes #878

### CODE REVIEW

File: src/zcl_excel_font.clas.abap [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    DATA ld_uccp                      TYPE i.
+    DATA ld_offset                    TYPE i.
+    DATA ld_width_from_font_metrics   TYPE i.
+    DATA ld_current_character         TYPE c LENGTH 1.
+    DATA ld_length                    TYPE i.
+    DATA ls_font_metric               TYPE mty_s_font_metric.
+
+    FIELD-SYMBOLS <ls_font_cache>  TYPE mty_s_font_cache.
+    FIELD-SYMBOLS <ls_itcfc>       TYPE itcfc.
+    FIELD-SYMBOLS <ls_font_metric> TYPE mty_s_font_metric.
```
Comment (sandraros) [inline src/zcl_excel_font.clas.abap]: When some code is moved, I check by comparing "manually" using notepad++ hence it's important to keep the original order and formatting.
```suggestion
    DATA: ld_cell_value                TYPE zexcel_cell_value,
          ld_style_guid                TYPE zexcel_cell_style,
          ls_stylemapping              TYPE zexcel_s_stylemapping,
          lo_table_object              TYPE REF TO object,
          lo_table                     TYPE REF TO zcl_excel_table,
          ld_table_top_left_column     TYPE zexcel_cell_column,
          ld_table_bottom_right_column TYPE zexcel_cell_column,
          ld_flag_contains_auto_filter TYPE abap_bool VALUE abap_false,
          ld_flag_bold                 TYPE abap_bool VALUE abap_false,
          ld_flag_italic               TYPE abap_bool VALUE abap_false,
          ld_date                      TYPE d,
          ld_date_char                 TYPE c LENGTH 50,
          ld_font_height               TYPE tdfontsize VALUE zcl_excel_font=>lc_default_font_height,
          ld_font_name                 TYPE zexcel_style_font_name VALUE zcl_excel_font=>lc_default_font_name.
```

File: src/zcl_excel_font.clas.abap [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+  PROTECTED SECTION.
+  PRIVATE SECTION.
+ENDCLASS.
+
+
+
+CLASS ZCL_EXCEL_FONT IMPLEMENTATION.
+
+
+  METHOD calculate.
```
Comment (sandraros) [inline src/zcl_excel_font.clas.abap]:
```suggestion
  METHOD calculate_text_width.
```

File: src/zcl_excel_font.clas.abap [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    TYPES:
+      mty_th_font_cache
+               TYPE HASHED TABLE OF mty_s_font_cache
+               WITH UNIQUE KEY font_name font_height flag_bold flag_italic .
+
+    CONSTANTS lc_default_font_height TYPE tdfontsize VALUE '110' ##NO_TEXT.
+    CONSTANTS lc_default_font_name TYPE zexcel_style_font_name VALUE 'Calibri' ##NO_TEXT.
+    CLASS-DATA mth_font_cache TYPE mty_th_font_cache .
+
+    CLASS-METHODS calculate
```
Comment (sandraros) [inline src/zcl_excel_font.clas.abap]:
```suggestion
    CLASS-METHODS calculate_text_width
```

File: src/zcl_excel_font.clas.testclasses.abap [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ENDCLASS.
+
+
+CLASS ltcl_Test IMPLEMENTATION.
+
+  METHOD calculate.
+
+    DATA lv_width TYPE f.
+
+    lv_width = zcl_excel_font=>calculate(
```
Comment (sandraros) [inline src/zcl_excel_font.clas.testclasses.abap]:
```suggestion
    lv_width = zcl_excel_font=>calculate_text_width(
```

File: src/zcl_excel_worksheet.clas.abap [hunk 5]
```diff
@@ -1340,116 +1296,12 @@ CLASS ZCL_EXCEL_WORKSHEET IMPLEMENTATION.
...
-          ADD ld_font_height TO ld_width_from_font_metrics.
-        ENDIF.
-      ENDDO.
-
-      " Add cell padding (Excel makes columns a bit wider than the space
-      " that is needed for the text itself) and convert unit
-      " (division by 100)
-      ep_width = ld_width_from_font_metrics / 100 + lc_excel_cell_padding.
-    ENDIF.
+    ep_width = zcl_excel_font=>calculate(
```
Comment (sandraros) [inline src/zcl_excel_worksheet.clas.abap]:
```suggestion
    ep_width = zcl_excel_font=>calculate_text_width(
```

**Review by sandraros — CHANGES_REQUESTED** [review]
Proposing to rename method (to indicate what exactly is calculated about font), and reorder/reformat data declarations in case in the future we need to compare what has been changed exactly.

File: src/zcl_excel_font.clas.abap [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    DATA ls_font_metric TYPE mty_s_font_metric.
+    DATA ld_length TYPE i.
+    DATA ld_offset TYPE i.
+    DATA ld_current_character TYPE c LENGTH 1.
+    DATA ld_width_from_font_metrics TYPE i.
+
+
+    FIELD-SYMBOLS <ls_font_cache>  TYPE mty_s_font_cache.
+    FIELD-SYMBOLS <ls_itcfc>       TYPE itcfc.
+    FIELD-SYMBOLS <ls_font_metric> TYPE mty_s_font_metric.
```
Comment (sandraros) [inline src/zcl_excel_font.clas.abap]: I don't know what happened to this proposal, but it's now worse. Now the code declares unused variables. Do I incorrectly propose the suggestion? Proposing again the same code.
```suggestion
    DATA: ld_current_character         TYPE c LENGTH 1,
          lt_itcfc                     TYPE STANDARD TABLE OF itcfc,
          ld_offset                    TYPE i,
          ld_length                    TYPE i,
          ld_uccp                      TYPE i,
          ls_font_metric               TYPE mty_s_font_metric,
          ld_width_from_font_metrics   TYPE i,
          ld_font_family               TYPE itcfh-tdfamily,
          lt_font_families             LIKE STANDARD TABLE OF ld_font_family,
          ls_font_cache                TYPE mty_s_font_cache.

    FIELD-SYMBOLS: <ls_font_cache>  TYPE mty_s_font_cache,
                   <ls_font_metric> TYPE mty_s_font_metric,
                   <ls_itcfc>       TYPE itcfc.
```

**Review by sandraros — CHANGES_REQUESTED** [review]
It seems there was a little problem when applying the suggestion, now there are unused variables. Suggesting the same proposal.

**Comment by larshp** [comment]
updated with proposal

**Review by sandraros — APPROVED** [review]
No regression on `ZDEMO_EXCEL31`. LGTM 👍

**Pull request merged**